File: elixir/3174_clear_digits.exs
# https://leetcode.com/problems/clear-digits/
# 10/02/2025

defmodule Solution do
  @spec clear_digits(String.t()) :: String.t()
  def clear_digits(s), do: clear_digits(s, [])

  defp clear_digits(<<>>, stack), do: stack |> Enum.reverse() |> List.to_string()

  defp clear_digits(<<c::utf8, s_tail::binary>>, [_ | stack_tail]) when c in ?0..?9,
    do: clear_digits(s_tail, stack_tail)

  defp clear_digits(<<c::utf8, s_tail::binary>>, stack), do: clear_digits(s_tail, [c | stack])
end

defmodule Main do
  def main do
    test_cases = [
      {
        "abc",
        "abc"
      },
      {
        "cb34",
        ""
      }
    ]

    test_cases
    |> Enum.each(fn {s, expected} ->
      result = Solution.clear_digits(s)
      IO.puts("Input: #{inspect(s)}")
      IO.puts("Output: #{inspect(result)}")
      IO.puts("Expected: #{inspect(expected)}")
      IO.puts("Result: #{if result == expected, do: "PASSED", else: "FAILED"}")
      IO.puts("------")
    end)
  end
end

Main.main()

File: rust/src/bin/3174_clear_digits.rs
// https://leetcode.com/problems/clear-digits/
// 10/02/2025

impl Solution {
    pub fn clear_digits(s: String) -> String {
        s.bytes()
            .fold(String::with_capacity(s.len()), |mut acc, c| {
                match c {
                    b'0'..=b'9' => _ = acc.pop(),
                    _ => acc.push(c as char),
                }

                acc
            })
    }
}

#[derive(Debug)]
struct Solution {}

#[cfg(test)]
mod tests {
    use super::*;
    use rstest::rstest;

    #[rstest]
    #[case("abc".to_string(), "abc".to_string())]
    #[case("cb34".to_string(), "".to_string())]
    fn test(#[case] s: String, #[case] expected: String) {
        let result = Solution::clear_digits(s);
        assert_eq!(result, expected);
    }
}

#[cfg(not(test))]
fn main() {
    let test_cases = vec![
        ("abc".to_string(), "abc".to_string()),
        ("cb34".to_string(), "".to_string()),
    ];

    for (s, expected) in test_cases {
        let result = Solution::clear_digits(s.clone());
        println!(
            "Input: {:?}, Expected: {:?}, Got: {:?}",
            s, expected, result
        );
        assert_eq!(result, expected);
    }
}
